### PR TITLE
GEOMESA-780 Kafka data store live playback WIP

### PR DESCRIPTION
**Pull request by mmatz-ccri**
* changed message protocol to include timestamp

### CODE REVIEW

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+sealed trait KafkaGeoMessage {
+
+  def timestamp: Instant
+}
+
+object KafkaGeoMessage {
+
+  implicit val clock = SystemClock
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: We usually use joda ```DateTime.now().withTimeZone(UTC)``` to explicitly set UTC time.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+/** Delete all [[SimpleFeature]]s
+  */
+case class Clear(override val timestamp: Instant) extends KafkaGeoMessage
+
+trait Clock {
+  def now: Instant
+}
+
+object SystemClock extends Clock {
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: Oh I see.  This seems unnecessary . . . maybe flatten this out and just use joda directly.

Comment (mmatz-ccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: The purpose is for testability - being able to use a known time when testing but I will flatten if you want.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+class KafkaGeoMessageDecoder(val schema: SimpleFeatureType) extends Logging {
+
+  type MSG = MessageAndMetadata[Array[Byte], Array[Byte]]
+
+  private lazy val V0_DELETE_KEY = "delete".getBytes(StandardCharsets.UTF_8)
+  private lazy val V0_CLEAR_KEY = "clear".getBytes(StandardCharsets.UTF_8)
+  private lazy val V0_TIMESTAMP = new Instant(0L)
+
+  val sfDecoder: SimpleFeatureDecoder = new AvroFeatureDecoder(schema, EncodingOptions.withUserData)
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: Should we default to using the Kryo encoder/decoder?

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        val id = new String(msg.message(), StandardCharsets.UTF_8)
+        Delete(ts, id)
+      case KafkaGeoMessageEncoder.clearType =>
+        Clear(ts)
+      case _ =>
+        throw new IllegalArgumentException("Unknown message type: " + msgType.toChar)
+    }
+  }
+
+  def decodeVersion0(msg: MSG): KafkaGeoMessage = {
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: I don't think there's reason to keep backwards compatibility at this point . . . but keep the version in the meta data for future backwards compatibility that *is* necessary.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala (hunk 2 above)
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: Oh, hmm.  Is there a way to work around that in the tests?  I hate to add layers of complexity *just* for testing.

**Comment by elahrvivaz** [comment]
looks good

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+sealed trait KafkaGeoMessage {
+
+  def timestamp: Instant
+}
+
+object KafkaGeoMessage {
+
+  def clear(): Clear =
+    new Clear(Instant.now)
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: Don't use ```new``` with case classes

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+  * The current version is 1
+  * The type is 'C' for create or update, 'D' for delete and 'X' for clear
+  *
+  * Value
+  *   CreateOrUpdate: serialized simple feature
+  *   Delete: simple feature ID as UTF-8 bytes
+  *   Clear: empty
+  *
+  */
+object KafkaGeoMessageEncoder {
```
Comment (anthonyccri) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessage.scala]: At some point, I'd like to divorce the message protocol from the transport mechanism.  I suggest renaming the messaging protocol classes and helper classes by dropping the 'Kafka' prefix.

**Pull request closed without merging**